File: leetcode/intersection_of_two_linked_lists.js
/* Write a program to find the node at which the intersection of 
 * two singly linked lists begins.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    for (var pa = headA, pb = headB; pa != pb;) {
        pa = pa ? pa.next : headB;
        pb = pb ? pb.next : headA;
    }
    return pa;
};

File: lintcode/intersection_of_two_linked_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param headA: the first list
    # @param headB: the second list
    # @return: a ListNode
    def getIntersectionNode(self, headA, headB):
        pa, pb = headA, headB
        while pa is not pb:
            pa = pa.next if pa is not None else headB
            pb = pb.next if pb is not None else headA
        return pa
